core-ktx/src/main/java/com/amplifyframework/kotlin/Storage.kt

```diff
@@ -0,0 +1,118 @@
+import com.amplifyframework.storage.result.StorageDownloadFileResult
+import com.amplifyframework.storage.result.StorageGetUrlResult
+import com.amplifyframework.storage.result.StorageListResult
+import com.amplifyframework.storage.result.StorageRemoveResult
+import com.amplifyframework.storage.result.StorageTransferProgress
+import com.amplifyframework.storage.result.StorageUploadFileResult
+import com.amplifyframework.storage.result.StorageUploadInputStreamResult
+import java.io.File
+import java.io.InputStream
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.FlowPreview
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.flattenMerge
+import kotlinx.coroutines.flow.flowOf
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.onEach
+
+@Suppress("unused")
+interface Storage {
+    @Throws(StorageException::class)
+    suspend fun getUrl(
+        key: String,
+        options: StorageGetUrlOptions = StorageGetUrlOptions.defaultInstance()
+    ):
+        StorageGetUrlResult
+
+    @ExperimentalCoroutinesApi
+    @FlowPreview
+    fun downloadFile(
+        key: String,
+        local: File,
+        options: StorageDownloadFileOptions = StorageDownloadFileOptions.defaultInstance()
+    ):
+        InProgressStorageOperation<StorageDownloadFileResult>
+
+    @ExperimentalCoroutinesApi
+    @FlowPreview
+    fun uploadFile(
+        key: String,
+        local: File,
+        options: StorageUploadFileOptions = StorageUploadFileOptions.defaultInstance()
+    ):
+        InProgressStorageOperation<StorageUploadFileResult>
+
+    @FlowPreview
+    @ExperimentalCoroutinesApi
+    fun uploadInputStream(
+        key: String,
+        local: InputStream,
+        options: StorageUploadInputStreamOptions = StorageUploadInputStreamOptions.defaultInstance()
+    ): InProgressStorageOperation<StorageUploadInputStreamResult>
+
+    @Throws(StorageException::class)
+    suspend fun remove(
+        key: String,
+        options: StorageRemoveOptions = StorageRemoveOptions.defaultInstance()
+    ): StorageRemoveResult
+
+    @Throws(StorageException::class)
+    suspend fun list(
+        path: String,
+        options: StorageListOptions = StorageListOptions.defaultInstance()
+    ): StorageListResult
+
+    @FlowPreview
+    data class InProgressStorageOperation<T>(
+        private val results: Flow<T>,
+        private val progress: Flow<StorageTransferProgress>,
+        private val errors: Flow<StorageException>,
+        private val delegate: Cancelable?
+    ) : Cancelable {
+
+        override fun cancel() {
+            delegate?.cancel()
+        }
+
+        @Suppress("UNCHECKED_CAST")
+        suspend fun result(): T {
+            // We want to take the first item from either one,
+            // without waiting for the other.
+            // Maybe there's a cleaner way to achieve this.
```

Comment:
Instead of having two `Flow`'s, (`results` and `errors`), could you just have `results`, and in the facade implementation emit success, as an item, or failure, as an error, directly to `results`?

Then, make `result` public, and you don't need this method at all any more.

Reply:
The Rx Bindings and Combine APIs work this way, today. The way you're suggesting is also a very reasonable mechanism. We can go the route you're suggesting, I'm open to it! It will mean changing all of our `Flow` APIs in this PR to use a single `Flow`, emitting `sealed class`es. This will ultimately work differently from any other API in iOS, Flutter, or Android -- but it might be the most ergonomic for the user?

If I understand you correctly, you're suggesting we do something like this?
```kotlin
Amplify.Storage.download(remoteKey, localFile)
    .collect {
        when (it) {
            is ProgressUpdate -> {
            },
            is TransferComplete -> {
            },
            is Error -> {
            }
    }

sealed class StorageEvent {
    data class ProgressUpdate(val increment, val total): StorageEvent
    data class TransferComplete<T>(result: T): StorageEvent
    data class Error(cause: StorageException): StorageEvent
}
```

Reply:
I haven't done much with `Flow`s so take this with a grain of salt. It's mostly academic.

With regards to errors, one approach could be to allow customers to use the [catch](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/catch.html) operator of the `Flow`. Unless we're doing any other processing of that error internally, I don't know that it makes sense to emit a special value for errors. So I think you would end up with something like this:

```kotlin
Amplify.Storage.download(remoteKey, localFile)
    .collect {
        when (it) {
            is ProgressUpdate -> {
            },
            is TransferComplete -> {
            }
    }
    .catch { storageExceptioon -> /* do whatever */ }
```

For progress updates, the sealed class approach seems solid. Only other approach I could think of, would be to make `InProgressStorageOperation<T>` a child class `MutableSharedFlow` and internally have the separate flow for progress updates. Invocations might look something like this:

```kotlin
Amplify.Storage.download(remoteKey, localFile)
    .collect { }
    .onProgress { }
    .catch { storageExceptioon -> /* do whatever */ }
```

I do understand the concern with keeping it consistent with the other platforms, but we should also keep in mind what approach feels more idiomatic for each of the dev stacks.

Reply:
FWIW, the current code should facilitate use of `.catch { }` as you mention. DataStore query & observe and API subscribe will currently emit errors like this. Storage upload/download will throw the error from the `result()` suspending function (it's like a single valued `Flow`, I suppose), but the `progress()` Flow never throws.

Reply:
Ah...got itt. So they would do something like this, correct?

```
Amplify.Storage.download(remoteKey, localFile)
    .results
    .collect {
        when (it) {
            is ProgressUpdate -> {
            },
            is TransferComplete -> {
            }
    }
    .catch { storageExceptioon -> /* do whatever */ }
```

Reply:
The download result is the odd-one-out. It works like a regular suspending function, e.g.:
```kotlin
val download = Amplify.Storage.downloadFile(remoteKey, localFile)
try {
    val result = download.result()
} catch (error: StorageException) {
    Log.w("DemoApp", "Failed to complete download", error)
}
```

The API subscribe has two `Flow`s in the operation. It *does* work the way you suggested:
```kotlin
Amplify.API.subscribe(request)
    .subscriptionData
    .catch { Log.w("DemoApp", "Error ended subscription", it) }
    .collect { Log.i("DemoApp", "New subscription event! $it") }
```